File: src/scopemeasure.cpp
#include "scopemeasure.h"

namespace Newsboat {

ScopeMeasure::ScopeMeasure(const std::string& func)
	: rs_object(scopemeasure::bridged::create(func))
{
}

void ScopeMeasure::stopover(const std::string& son)
{
	scopemeasure::bridged::stopover(*rs_object, son);
}

} // namespace Newsboat

File: rust/libnewsboat-ffi/src/scopemeasure.rs
use libNewsboat::scopemeasure;

// cxx doesn't allow to share types from other crates, so we have to wrap it
// cf. https://github.com/dtolnay/cxx/issues/496
struct ScopeMeasure(scopemeasure::ScopeMeasure);

#[cxx::bridge(namespace = "Newsboat::scopemeasure::bridged")]
mod bridged {
    extern "Rust" {
        type ScopeMeasure;

        fn create(scope_name: String) -> Box<ScopeMeasure>;
        fn stopover(obj: &ScopeMeasure, stopover_name: &str);
    }
}

fn create(scope_name: String) -> Box<ScopeMeasure> {
    Box::new(ScopeMeasure(scopemeasure::ScopeMeasure::new(scope_name)))
}

fn stopover(obj: &ScopeMeasure, stopover_name: &str) {
    obj.0.stopover(stopover_name);
}
